File: modal-rb/lib/modal/sandbox_filesystem.rb
module Modal
  class SandboxFile
    attr_reader :file_descriptor, :task_id

    def initialize(file_descriptor, task_id)
      @file_descriptor = file_descriptor
      @task_id = task_id
    end

    def read
      request = Modal::Client::ContainerFilesystemExecRequest.new(
        file_read_request: Modal::Client::ContainerFileReadRequest.new(
          file_descriptor: @file_descriptor
        ),
        task_id: @task_id
      )

      resp = Modal.client.call(:container_filesystem_exec, request)
      exec_id = resp.exec_id

      data = ""
      completed = false
      retries = 20

      while !completed && retries > 0
        begin
          output_request = Modal::Client::ContainerFilesystemExecGetOutputRequest.new(
            exec_id: exec_id,
            timeout: 10
          )

          stream = Modal.client.call(:container_filesystem_exec_get_output, output_request)

          stream.each do |batch|

            if batch.respond_to?(:output) && batch.output && batch.output.any?
              if batch.output.first.is_a?(String)
                chunk = batch.output.join('')
              elsif batch.output.first.is_a?(Integer)
                chunk = batch.output.pack('C*')
              else
                chunk = batch.output.map(&:to_s).join('')
              end

              data += chunk.force_encoding('UTF-8')
            end

            if batch.respond_to?(:error) && batch.error
              raise SandboxFilesystemError.new("Read failed: #{batch.error.error_message}")
            end

            if batch.respond_to?(:eof) && batch.eof
              completed = true
              break
            end
          end

          unless completed
            retries -= 1
            sleep(0.1)
          end

        rescue GRPC::BadStatus => e
          if e.code == GRPC::Core::StatusCodes::DEADLINE_EXCEEDED
            retries -= 1
            sleep(0.1)
            next
          else
            raise e
          end
        end
      end

      data
    end

    def write(data)
      binary_data = if data.is_a?(String)
                      data.force_encoding('BINARY')
                    else
                      data.to_s.force_encoding('BINARY')
                    end

      request = Modal::Client::ContainerFilesystemExecRequest.new(
        file_write_request: Modal::Client::ContainerFileWriteRequest.new(
          file_descriptor: @file_descriptor,
          data: binary_data
        ),
        task_id: @task_id
      )


      resp = Modal.client.call(:container_filesystem_exec, request)
      exec_id = resp.exec_id


      completed = false
      retries = 20

      while !completed && retries > 0
        begin
          output_request = Modal::Client::ContainerFilesystemExecGetOutputRequest.new(
            exec_id: exec_id,
            timeout: 10
          )

          stream = Modal.client.call(:container_filesystem_exec_get_output, output_request)

          stream.each do |batch|
            if batch.respond_to?(:error) && batch.error
              raise SandboxFilesystemError.new("Write failed: #{batch.error.error_message}")
            end
            if batch.respond_to?(:eof) && batch.eof
              completed = true
              break
            end
          end

          unless completed
            retries -= 1
            sleep(0.1)
          end

        rescue GRPC::BadStatus => e
          if e.code == GRPC::Core::StatusCodes::DEADLINE_EXCEEDED
            retries -= 1
            sleep(0.1)
            next
          else
            raise e
          end
        end
      end

      unless completed
      end

      data.length
    end

    def flush
      request = Modal::Client::ContainerFilesystemExecRequest.new(
        file_flush_request: Modal::Client::ContainerFileFlushRequest.new(
          file_descriptor: @file_descriptor
        ),
        task_id: @task_id
      )

      resp = Modal.client.call(:container_filesystem_exec, request)
      exec_id = resp.exec_id

      retries = 10

      while retries > 0
        begin
          output_request = Modal::Client::ContainerFilesystemExecGetOutputRequest.new(
            exec_id: exec_id,
            timeout: 5
          )

          stream = Modal.client.call(:container_filesystem_exec_get_output, output_request)
          stream.each do |batch|
            if batch.respond_to?(:error) && batch.error
              raise SandboxFilesystemError.new("Flush failed: #{batch.error.error_message}")
            end
            if batch.respond_to?(:eof) && batch.eof
              return
            end
          end

          retries -= 1
          sleep(0.1)

        rescue GRPC::BadStatus => e
          if e.code == GRPC::Core::StatusCodes::DEADLINE_EXCEEDED
            retries -= 1
            next
          else
            raise e
          end
        end
      end
    end

    def close
      request = Modal::Client::ContainerFilesystemExecRequest.new(
        file_close_request: Modal::Client::ContainerFileCloseRequest.new(
          file_descriptor: @file_descriptor
        ),
        task_id: @task_id
      )

      resp = Modal.client.call(:container_filesystem_exec, request)
      exec_id = resp.exec_id

      retries = 10

      while retries > 0
        begin
          output_request = Modal::Client::ContainerFilesystemExecGetOutputRequest.new(
            exec_id: exec_id,
            timeout: 5
          )

          stream = Modal.client.call(:container_filesystem_exec_get_output, output_request)
          stream.each do |batch|
            if batch.respond_to?(:error) && batch.error
              raise SandboxFilesystemError.new("Close failed: #{batch.error.error_message}")
            end
            if batch.respond_to?(:eof) && batch.eof
              return
            end
          end

          retries -= 1
          sleep(0.1)

        rescue GRPC::BadStatus => e
          if e.code == GRPC::Core::StatusCodes::DEADLINE_EXCEEDED
            retries -= 1
            next
          else
            raise e
          end
        end
      end
    end
  end
end

File: modal-kt/src/main/kotlin/com/modal/modalkt/SandboxFilesystem.kt
package com.modal.modalkt

import kotlinx.coroutines.flow.collect
import modal.client.Api
import java.io.ByteArrayOutputStream

typealias SandboxFileMode = String

class SandboxFile(
    private val client: ModalClient,
    private val fileDescriptor: String,
    private val taskId: String,
) {
    suspend fun read(): ByteArray {
        val result = runFilesystemExec(
            client.cpClient,
            Api.ContainerFilesystemExecRequest.newBuilder()
                .setTaskId(taskId)
                .setFileReadRequest(
                    Api.ContainerFileReadRequest.newBuilder()
                        .setFileDescriptor(fileDescriptor)
                        .build(),
                )
                .build(),
        )
        return result.chunks
    }

    suspend fun write(data: ByteArray) {
        runFilesystemExec(
            client.cpClient,
            Api.ContainerFilesystemExecRequest.newBuilder()
                .setTaskId(taskId)
                .setFileWriteRequest(
                    Api.ContainerFileWriteRequest.newBuilder()
                        .setFileDescriptor(fileDescriptor)
                        .setData(com.google.protobuf.ByteString.copyFrom(data))
                        .build(),
                )
                .build(),
        )
    }

    suspend fun flush() {
        runFilesystemExec(
            client.cpClient,
            Api.ContainerFilesystemExecRequest.newBuilder()
                .setTaskId(taskId)
                .setFileFlushRequest(
                    Api.ContainerFileFlushRequest.newBuilder()
                        .setFileDescriptor(fileDescriptor)
                        .build(),
                )
                .build(),
        )
    }

    suspend fun close() {
        runFilesystemExec(
            client.cpClient,
            Api.ContainerFilesystemExecRequest.newBuilder()
                .setTaskId(taskId)
                .setFileCloseRequest(
                    Api.ContainerFileCloseRequest.newBuilder()
                        .setFileDescriptor(fileDescriptor)
                        .build(),
                )
                .build(),
        )
    }
}

suspend fun runFilesystemExec(
    cpClient: ControlPlaneClient,
    request: Api.ContainerFilesystemExecRequest,
): FilesystemExecResult {
    val response = cpClient.containerFilesystemExec(request)
    val output = ByteArrayOutputStream()
    var retries = 10
    var completed = false

    while (!completed) {
        try {
            cpClient.containerFilesystemExecGetOutput(
                Api.ContainerFilesystemExecGetOutputRequest.newBuilder()
                    .setExecId(response.execId)
                    .setTimeout(55f)
                    .build(),
            ).collect { batch ->
                for (chunk in batch.outputList) {
                    output.write(chunk.toByteArray())
                }
                if (batch.hasError()) {
                    if (retries > 0) {
                        retries -= 1
                        return@collect
                    }
                    throw SandboxFilesystemError(batch.error.errorMessage)
                }
                if (batch.eof) {
                    completed = true
                }
            }
        } catch (error: Throwable) {
            if (isRetryableGrpc(error) && retries > 0) {
                retries -= 1
            } else {
                throw error
            }
        }
    }

    return FilesystemExecResult(output.toByteArray(), response)
}

data class FilesystemExecResult(
    val chunks: ByteArray,
    val response: Api.ContainerFilesystemExecResponse,
)
